{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fbaf7a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20f0704f",
   "metadata": {},
   "source": [
    "# the list of housekeeping genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9b753a3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABHD5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALDH3A2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene\n",
       "0     TP63\n",
       "1   ABCA12\n",
       "2    ABHD5\n",
       "3     AGPS\n",
       "4  ALDH3A2"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the list of housekeeping genes:\n",
    "hk_genes = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\Heatmap_gene_Expression\\housekeeping_csv.csv\")\n",
    "hk_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8b4a41",
   "metadata": {},
   "source": [
    "# adding target gene to the list of housekeeping genes:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ccae10ff",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\f.vahidnejhad\\AppData\\Local\\Temp\\ipykernel_19708\\3854047640.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  hk_genes = hk_genes.append(gene_target, ignore_index = True)\n"
     ]
    }
   ],
   "source": [
    "#adding target gene to the list of housekeeping genes:\n",
    "gene_name = ['TP63']\n",
    "gene_target = pd.DataFrame(gene_name, columns = ['Gene'])\n",
    "gene_target\n",
    "hk_genes = hk_genes.append(gene_target, ignore_index = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cfc693e",
   "metadata": {},
   "source": [
    "# importing samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "000c07fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample 1\n",
    "spl_1 = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\csv_format\\ZaKh-L_S.csv\",low_memory=False)\n",
    "\n",
    "spl_1 = spl_1.rename(columns={'Name': 'Gene', 'Expression value' : 'ZaKh-L_S'  })\n",
    "#spl_1.head()\n",
    "\n",
    "\n",
    "\n",
    "#smaple 2\n",
    "spl_2 = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\csv_format\\ZaKh-N_S8_L002_R1_001 (GE).csv\",low_memory=False)\n",
    "\n",
    "spl_2 = spl_2.rename(columns={'Name': 'Gene' , 'Expression value' : 'ZaKh-N_S8'})\n",
    "#spl_2.head()\n",
    "\n",
    "\n",
    "\n",
    "#sample 3\n",
    "spl_3 = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\csv_format\\AL-Kh-N_S6_L001_R1_001 (GE).csv\",low_memory=False)\n",
    "\n",
    "spl_3 = spl_3.rename(columns={'Name': 'Gene' , 'Expression value' : 'AL-Kh-N_S6'})\n",
    "#spl_3.head()\n",
    "\n",
    "\n",
    "\n",
    "#sample 4\n",
    "spl_4 = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\csv_format\\Ar-Kh-W_S18_L001_R1_001 (GE).csv\",low_memory=False)\n",
    "\n",
    "spl_4 = spl_4.rename(columns={'Name': 'Gene', 'Expression value' : 'Ar-Kh-W_S18'})\n",
    "#spl_4.head()\n",
    "\n",
    "\n",
    "\n",
    "#smaple 5\n",
    "ctrlz = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\csv_format\\ControlZ1(GE)-failed.csv\",low_memory=False)\n",
    "\n",
    "ctrlz = ctrlz.rename(columns={'Name': 'Gene' , 'Expression value' : 'ctrlz'})\n",
    "#ctrlz.head()\n",
    "\n",
    "\n",
    "\n",
    "#smaple 6\n",
    "ctrlb = pd.read_csv(r\"C:\\Users\\f.vahidnejhad\\Desktop\\private\\TP63\\csv_format\\ControlB1(GE).csv\",low_memory=False)\n",
    "\n",
    "ctrlb = ctrlb.rename(columns={'Name': 'Gene' , 'Expression value' : 'ctrlb'})\n",
    "#ctrlb.head()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f76c3cf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>ctrlb</th>\n",
       "      <th>TPM</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>Exons</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene length</th>\n",
       "      <th>ENSEMBL</th>\n",
       "      <th>...</th>\n",
       "      <th>Uniquely identified transcripts</th>\n",
       "      <th>Exon length</th>\n",
       "      <th>Unique exon reads</th>\n",
       "      <th>Total exon reads</th>\n",
       "      <th>Ratio of unique to total (exon reads)</th>\n",
       "      <th>Unique exon-exon reads</th>\n",
       "      <th>Total exon-exon reads</th>\n",
       "      <th>Unique intron reads</th>\n",
       "      <th>Total intron reads</th>\n",
       "      <th>Ratio of intron to total gene reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>65419..71585</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>6167</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2618</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>#NUM!</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>#NUM!</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>1</td>\n",
       "      <td>complement(450703..451697)</td>\n",
       "      <td>2</td>\n",
       "      <td>0.238954</td>\n",
       "      <td>0.024870</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000284733</td>\n",
       "      <td>995</td>\n",
       "      <td>ENSG00000284733</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>995</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>1</td>\n",
       "      <td>complement(685679..686673)</td>\n",
       "      <td>2</td>\n",
       "      <td>0.238954</td>\n",
       "      <td>0.024870</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>995</td>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>995</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>923928..944581</td>\n",
       "      <td>19</td>\n",
       "      <td>0.606041</td>\n",
       "      <td>0.063076</td>\n",
       "      <td>29</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>20654</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3727</td>\n",
       "      <td>17</td>\n",
       "      <td>19</td>\n",
       "      <td>0.894736842</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>0.457142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1</td>\n",
       "      <td>complement(944203..959309)</td>\n",
       "      <td>1868</td>\n",
       "      <td>80.546710</td>\n",
       "      <td>8.383256</td>\n",
       "      <td>19</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>15107</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2757</td>\n",
       "      <td>1672</td>\n",
       "      <td>1868</td>\n",
       "      <td>0.895074946</td>\n",
       "      <td>1342</td>\n",
       "      <td>1342</td>\n",
       "      <td>245</td>\n",
       "      <td>245</td>\n",
       "      <td>0.115948888</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene Chromosome                      Region  ctrlb        TPM      RPKM  \\\n",
       "0   OR4F5          1                65419..71585      0   0.000000  0.000000   \n",
       "1  OR4F29          1  complement(450703..451697)      2   0.238954  0.024870   \n",
       "2  OR4F16          1  complement(685679..686673)      2   0.238954  0.024870   \n",
       "3  SAMD11          1              923928..944581     19   0.606041  0.063076   \n",
       "4   NOC2L          1  complement(944203..959309)   1868  80.546710  8.383256   \n",
       "\n",
       "   Exons          Gene ID  Gene length          ENSEMBL  ...  \\\n",
       "0      4  ENSG00000186092         6167  ENSG00000186092  ...   \n",
       "1      1  ENSG00000284733          995  ENSG00000284733  ...   \n",
       "2      1  ENSG00000284662          995  ENSG00000284662  ...   \n",
       "3     29  ENSG00000187634        20654  ENSG00000187634  ...   \n",
       "4     19  ENSG00000188976        15107  ENSG00000188976  ...   \n",
       "\n",
       "  Uniquely identified transcripts  Exon length  Unique exon reads  \\\n",
       "0                               0         2618                  0   \n",
       "1                               0          995                  0   \n",
       "2                               0          995                  0   \n",
       "3                               1         3727                 17   \n",
       "4                               1         2757               1672   \n",
       "\n",
       "   Total exon reads  Ratio of unique to total (exon reads)  \\\n",
       "0                 0                                  #NUM!   \n",
       "1                 2                                      0   \n",
       "2                 2                                      0   \n",
       "3                19                            0.894736842   \n",
       "4              1868                            0.895074946   \n",
       "\n",
       "   Unique exon-exon reads  Total exon-exon reads  Unique intron reads  \\\n",
       "0                       0                      0                    0   \n",
       "1                       0                      0                    0   \n",
       "2                       0                      0                    0   \n",
       "3                       9                      9                   16   \n",
       "4                    1342                   1342                  245   \n",
       "\n",
       "  Total intron reads  Ratio of intron to total gene reads  \n",
       "0                  0                                #NUM!  \n",
       "1                  1                          0.333333333  \n",
       "2                  2                                  0.5  \n",
       "3                 16                          0.457142857  \n",
       "4                245                          0.115948888  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctrlb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "62dfb737",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>ZaKh-L_S</th>\n",
       "      <th>ZaKh-N_S8</th>\n",
       "      <th>AL-Kh-N_S6</th>\n",
       "      <th>Ar-Kh-W_S18</th>\n",
       "      <th>ctrlz</th>\n",
       "      <th>ctrlb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP63</td>\n",
       "      <td>9919.0</td>\n",
       "      <td>7917.0</td>\n",
       "      <td>7857.0</td>\n",
       "      <td>17664.0</td>\n",
       "      <td>10216.0</td>\n",
       "      <td>11966.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA12</td>\n",
       "      <td>2242.0</td>\n",
       "      <td>3791.0</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>6193.0</td>\n",
       "      <td>8017.0</td>\n",
       "      <td>7948.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABHD5</td>\n",
       "      <td>599.0</td>\n",
       "      <td>844.0</td>\n",
       "      <td>861.0</td>\n",
       "      <td>2752.0</td>\n",
       "      <td>3381.0</td>\n",
       "      <td>3618.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGPS</td>\n",
       "      <td>1227.0</td>\n",
       "      <td>723.0</td>\n",
       "      <td>478.0</td>\n",
       "      <td>1463.0</td>\n",
       "      <td>2990.0</td>\n",
       "      <td>2364.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALDH3A2</td>\n",
       "      <td>1828.0</td>\n",
       "      <td>5602.0</td>\n",
       "      <td>7928.0</td>\n",
       "      <td>6470.0</td>\n",
       "      <td>10780.0</td>\n",
       "      <td>11492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3604</th>\n",
       "      <td>NDUFC2-KCTD14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3605</th>\n",
       "      <td>NEDD8-MDP1</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3606</th>\n",
       "      <td>COMMD3-BMI1</td>\n",
       "      <td>469.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>136.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3607</th>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>597.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>437.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>523.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608</th>\n",
       "      <td>UTP14C</td>\n",
       "      <td>725.0</td>\n",
       "      <td>552.0</td>\n",
       "      <td>839.0</td>\n",
       "      <td>1075.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3583 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene  ZaKh-L_S  ZaKh-N_S8  AL-Kh-N_S6  Ar-Kh-W_S18    ctrlz  \\\n",
       "0              TP63    9919.0     7917.0      7857.0      17664.0  10216.0   \n",
       "1            ABCA12    2242.0     3791.0      3131.0       6193.0   8017.0   \n",
       "2             ABHD5     599.0      844.0       861.0       2752.0   3381.0   \n",
       "3              AGPS    1227.0      723.0       478.0       1463.0   2990.0   \n",
       "4           ALDH3A2    1828.0     5602.0      7928.0       6470.0  10780.0   \n",
       "...             ...       ...        ...         ...          ...      ...   \n",
       "3604  NDUFC2-KCTD14       0.0        0.0        27.0         54.0     48.0   \n",
       "3605     NEDD8-MDP1      22.0        0.0        18.0          8.0      3.0   \n",
       "3606    COMMD3-BMI1     469.0       65.0       142.0         84.0    151.0   \n",
       "3607         RSC1A1     597.0      190.0        18.0        437.0    406.0   \n",
       "3608         UTP14C     725.0      552.0       839.0       1075.0      0.0   \n",
       "\n",
       "        ctrlb  \n",
       "0     11966.0  \n",
       "1      7948.0  \n",
       "2      3618.0  \n",
       "3      2364.0  \n",
       "4     11492.0  \n",
       "...       ...  \n",
       "3604      9.0  \n",
       "3605      0.0  \n",
       "3606    136.0  \n",
       "3607    523.0  \n",
       "3608      0.0  \n",
       "\n",
       "[3583 rows x 7 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = hk_genes.merge(spl_1[['Gene', 'ZaKh-L_S']], \n",
    "                    on = 'Gene', how = 'left').merge(spl_2[['Gene','ZaKh-N_S8']],\n",
    "                    on = 'Gene', how = 'left').merge(spl_3[['Gene', 'AL-Kh-N_S6']],\n",
    "                    on = 'Gene', how = 'left').merge(spl_4[['Gene', 'Ar-Kh-W_S18']],\n",
    "                    on = 'Gene', how = 'left').merge(ctrlz[['Gene', 'ctrlz']],\n",
    "                     on = 'Gene', how = 'left').merge(ctrlb[['Gene', 'ctrlb']],\n",
    "                     on = 'Gene', how = 'left')\n",
    "\n",
    "df1 = df1.dropna(axis = 0)\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5df37a4",
   "metadata": {},
   "source": [
    "# Normalizing the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "7bd55fb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.5\n"
     ]
    }
   ],
   "source": [
    "def average (lst):\n",
    "    return sum(lst)/len(lst)\n",
    "\n",
    "\n",
    "\n",
    "def normalize_data (*args):\n",
    "    sum_expression = []\n",
    "    norm_fact = []\n",
    "    \n",
    "    #sumation of gene expression for every sample:\n",
    "    for arg in args:\n",
    "        a = arg.sum()\n",
    "        sum_expression.append(a)\n",
    "    \n",
    "    #average of gene experssion:\n",
    "    avg = average(sum_expression)\n",
    "    \n",
    "    \n",
    "    #norm_fact:\n",
    "    for i in sum_expression:\n",
    "        norm_fact.append(i/avg)\n",
    "        \n",
    "        \n",
    "    return sum_expression  , norm_fact , avg\n",
    "        \n",
    "       \n",
    "\n",
    "a = df1['ZaKh-L_S']\n",
    "b = df1['ZaKh-N_S8']\n",
    "c = df1['AL-Kh-N_S6']\n",
    "d = df1['Ar-Kh-W_S18']\n",
    "e = df1['ctrlb']\n",
    "f = df1['ctrlz']\n",
    "sum_expression  , norm_fact , avg = normalize_data(a,b,c,d,e,f)\n",
    "#print(sum_expression  , norm_fact , avg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "28a2c81c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ZaKh-L_S 0.5389647583863769\n",
      "ZaKh-N_S8 0.5611881018832381\n",
      "AL-Kh-N_S6 1.0703959452488232\n",
      "Ar-Kh-W_S18 1.0631361144010083\n",
      "ctrlb 1.3960331282704272\n",
      "ctrlz 1.3702819518101266\n"
     ]
    }
   ],
   "source": [
    "name_samples = ['ZaKh-L_S','ZaKh-N_S8','AL-Kh-N_S6','Ar-Kh-W_S18','ctrlb','ctrlz' ]\n",
    "\n",
    "\n",
    "dic = {k:v for k,v in zip(name_samples,norm_fact)}\n",
    "for key, value in dic.items():\n",
    "    print(key,value)\n",
    "    \n",
    "\n",
    "    \n",
    "new_df = df1.set_index('Gene')\n",
    "normalized_data = new_df.div(dic)\n",
    "normalized_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "baa0bd08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZaKh-L_S</th>\n",
       "      <th>ZaKh-N_S8</th>\n",
       "      <th>AL-Kh-N_S6</th>\n",
       "      <th>Ar-Kh-W_S18</th>\n",
       "      <th>ctrlz</th>\n",
       "      <th>ctrlb</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TP63</th>\n",
       "      <td>18403.800704</td>\n",
       "      <td>14107.569233</td>\n",
       "      <td>7340.274442</td>\n",
       "      <td>16614.993848</td>\n",
       "      <td>7455.399954</td>\n",
       "      <td>8571.42983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ZaKh-L_S     ZaKh-N_S8   AL-Kh-N_S6   Ar-Kh-W_S18        ctrlz  \\\n",
       "Gene                                                                       \n",
       "TP63  18403.800704  14107.569233  7340.274442  16614.993848  7455.399954   \n",
       "\n",
       "           ctrlb  \n",
       "Gene              \n",
       "TP63  8571.42983  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_data.loc[gene_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "1a35de52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#report the normalized dataset:\n",
    "normalized_data.to_csv('normalized_data.csv', index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
